### PR TITLE
Bug 1314227 - GPU process telemetry support for aggregator

### PR DESCRIPTION
**Pull request by chutten**
This is surprisingly less-complicated than I feared.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/python_mozaggregator/29)

### CODE REVIEW

**Comment by chutten** [comment]
Should obsolete #28 if that's okay with you, @gw280

File: tests/test_aggregator.py [hunk 1]
```diff
@@ -130,7 +130,7 @@ def test_classic_histograms():
 
     assert(len(metric_count) == len(histograms))
     for v in metric_count.values():
-        assert(v == 2*len(build_id_aggregates))  # Count both child and parent metrics
+        assert(v == (1 + NUM_CHILD_PROCESS_TYPES)*len(build_id_aggregates))  # Count both child and parent metrics
```
Comment (fbertsch) [inline tests/test_aggregator.py]: I think this should be updated. This is checking that the sum total we've seen is expected; but we never actually check that we've seen every expected process type. We should instead have a metric_count[metric][process_type], where the value of that should be len(build_id_aggregates). Then we can ensure we've seen all process types and that their values are correct.

Note that this should be changed in test_count_histograms, test_keyed_histograms, and test_user_counter2_histograms as well.

File: tests/test_db.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def test_filters():
             assert set(options["application"]) == set(ping_dimensions["application"])
             assert set(options["architecture"]) == set(ping_dimensions["arch"])
             assert set(options["e10sEnabled"]) == set(["true", "false"])
-            assert set(options["child"]) == set(["true", "false"])
+            assert set(options["child"]) == set(["gpu", "content", "parent"])
```
Comment (fbertsch) [inline tests/test_db.py]: Optional: Move this to a constant, set NUM_PROCESS_TYPES equal to it's length, and then use it to ensure the set of keys in metric_count[metric] (in test_classic_histograms and it's ilk, per my other comment) contain all processes.

File: mozaggregator/db.py [hunk 3]
```diff
@@ -132,7 +132,10 @@ def _aggregate_to_sql(aggregate):
 
         dimensions["metric"] = metric
         dimensions["label"] = label
-        dimensions["child"] = child
+        # Have to special-case content and parent here to maintain backwards compatibility.
+        dimensions["child"] = "true" if child == "content" \
```
Comment (fbertsch) [inline mozaggregator/db.py]: s/child/process_type. Please do this across all files.

Comment (chutten) [inline mozaggregator/db.py]: At this, the db level, don't we need it to remain "child" for backwards compat?

Comment (fbertsch) [inline mozaggregator/db.py]: Yes, sorry, I would have liked to comment on line 113, but github didn't let me :) I mean the variable names.

File: mozaggregator/db.py [hunk 4]
```diff
@@ -132,7 +132,10 @@ def _aggregate_to_sql(aggregate):
...
+                                else "false" if child == "parent" \
+                                  else child
```
Comment (chutten) [inline mozaggregator/db.py]: There are 629 `flake8` errors in this project. :S

I will make these indents adhere, but I think the other 625 may have to wait for another day

Comment (fbertsch) [inline mozaggregator/db.py]: We gotta start somewhere :)

**Comment by fbertsch** [comment]
Additionally, please combine these two commits into a single atomic commit. This will cut out gw280's commit, but is far easier to maintain.

**Comment by chutten** [comment]
Hey, whaddayaknow, git/github knows that gw280 and I share ownership of the squashed commit. Noice.

File: tests/dataset.py [hunk 5]
```diff
@@ -166,7 +167,11 @@ def generate_payload(dimensions, aggregated_child_histograms):
     if aggregated_child_histograms:
         processes_payload[u"content"] = {
             u"histograms": histograms_template,
-            u"keyedHistograms": keyed_histograms_template 
+            u"keyedHistograms": keyed_histograms_template
+        }
+        processes_payload[u"gpu"] = {
```
Comment (fbertsch) [inline tests/dataset.py]: I think bug 1278556 included support for gpu processes scalars. Please either include scalars, or amend the commit title to specify gpu histograms. (No problem if you do the latter, I need to add content processes scalars as well, so could do both of these in one pass).

**Comment by fbertsch** [comment]
Reviewed 1 of 3 files at r1, 3 of 5 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29)*

**Comment by fbertsch** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-:-Kc-bT46tlRTvINXwanD:bnfp4nl)*

**Comment by fbertsch** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[mozaggregator/db.py, line 136 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECtS9k-o3clgoaV-r2-136:-Kc-bY7qb4SSsKY41bmT:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/mozaggregator/db.py#L136)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Yes, sorry, I would have liked to comment on line 113, but github didn't let me :) I mean the variable names.
</blockquote></details>

OK

---

*[mozaggregator/db.py, line 138 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECtS9k-o3clgoaV-r2-138:-Kc-balAiynEFNh6E-pz:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/mozaggregator/db.py#L138)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

We gotta start somewhere :)
</blockquote></details>

OK

---

*[mozaggregator/service.py, line 132 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECtS9k-o3clgoaW-r2-132:-Kc-bbOi2IgJPeyJSxT1:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/mozaggregator/service.py#L132)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Please align these indents (check with flake8 if unsure)
</blockquote></details>

OK

---

*[tests/dataset.py, line 9 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECsGkKeYm1LjaBw-r2-9:-Kc-bc2AA6lXzu0t26XD:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/tests/dataset.py#L9)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Please change this to "NUM_PROCESS_TYPES = 3", and we can remove the "1 + "'s below
</blockquote></details>

OK

---

*[tests/dataset.py, line 204 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECsGkKeYm1LjaBw-r2-204:-Kc-bd6CBHGCJxt_T4N7:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/tests/dataset.py#L204)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Please check here that it is a content process.
</blockquote></details>

OK

---

*[tests/dataset.py, line 172 at r3](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECsGkKeYm1LjaBw-r3-172:-Kc-bc_y4YRU9Osd2D8Y:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/3d1e897f616482dcf48ebc54890cd7f7c3e8f085/tests/dataset.py#L172)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

I think bug 1278556 included support for gpu processes scalars. Please either include scalars, or amend the commit title to specify gpu histograms. (No problem if you do the latter, I need to add content processes scalars as well, so could do both of these in one pass).
</blockquote></details>

OK

---

*[tests/test_aggregator.py, line 133 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECtS9k-o3clgoaX-r2-133:-Kc-beTqukt0zpZhDR0d:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/tests/test_aggregator.py#L133)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

I think this should be updated. This is checking that the sum total we've seen is expected; but we never actually check that we've seen every expected process type. We should instead have a metric_count[metric][process_type], where the value of that should be len(build_id_aggregates). Then we can ensure we've seen all process types and that their values are correct.

Note that this should be changed in test_count_histograms, test_keyed_histograms, and test_user_counter2_histograms as well.
</blockquote></details>

OK

---

*[tests/test_aggregator.py, line 121 at r3](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECtS9k-o3clgoaX-r3-121:-Kc-bdf-CszSJa0axWGp:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/3d1e897f616482dcf48ebc54890cd7f7c3e8f085/tests/test_aggregator.py#L121)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Please make these `defaultdict(lambda: defaultdict(int))`, and then we can remove the `if metric not in metric_count` on line 130.
</blockquote></details>

OK

---

*[tests/test_aggregator.py, line 140 at r3](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECtS9k-o3clgoaX-r3-140:-Kc-be2ryUAD3bM97DdC:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/3d1e897f616482dcf48ebc54890cd7f7c3e8f085/tests/test_aggregator.py#L140)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Please assert that we have the correct number of process types.
</blockquote></details>

OK

---

*[tests/test_db.py, line 97 at r2](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECsGkKeYm1LjaBx-r2-97:-Kc-betRB3juUpcRs8Ib:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/678ca27fbe2952534d96a36135ebc5174f92f279/tests/test_db.py#L97)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Optional: Move this to a constant, set NUM_PROCESS_TYPES equal to it's length, and then use it to ensure the set of keys in metric_count[metric] (in test_classic_histograms and it's ilk, per my other comment) contain all processes.
</blockquote></details>

OK

---

*[tests/test_db.py, line 242 at r3](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29#-KbHBECsGkKeYm1LjaBx-r3-242:-Kc-bfHNDb-IviZiQBT5:b32w0x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/3d1e897f616482dcf48ebc54890cd7f7c3e8f085/tests/test_db.py#L242)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

I'm not sure this one needs to be changed. It's just used to check whether the associated measure type is a child payload. It's only relevant because simple measures are not aggregated under the "processes" portion of the ping.
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/29)*

**Comment by fbertsch** [comment]
Reviewable is terrible, and is not letting me merge this. @vitillo can you disable reviewable for this repo?

**Comment by vitillo** [comment]
Yep, I was just about to.

**Pull request merged**